R/estimatr_lh_robust.R

```diff
@@ -0,0 +1,161 @@
+#' following components:
+#'   \item{coefficients}{the estimated linear combination of coefficients}
+#'   \item{std.error}{the estimated standard errors of the linear combination}
+#'   \item{statistic}{the t-statistic}
+#'   \item{df}{the estimated degrees of freedom}
+#'   \item{p.value}{the p-values from a two-sided t-test using \code{coefficients}, \code{std.error}, and \code{df}}
+#'   \item{conf.low}{the lower bound of the \code{1 - alpha} percent confidence interval}
+#'   \item{conf.high}{the upper bound of the \code{1 - alpha} percent confidence interval}
+#'   \item{term}{a character vector of coefficient names}
+#'   \item{alpha}{the significance level specified by the user}
+#' @export
+#'
+lh_robust <- function(formula,
+                      data,
+                      weights,
+                      subset,
+                      clusters,
+                      fixed_effects,
+                      se_type = NULL,
+                      ci = TRUE,
+                      alpha = .05,
+                      return_vcov = TRUE,
+                      try_cholesky = FALSE,
+                      linearHypothesis = NULL,
+                      lh_term = NULL
+                      ) {
+
+  return_list <- list()
+
+
+  datargs <- enquos(
+    formula = formula,
+    weights = weights,
+    subset = subset,
+    cluster = clusters,
+    fixed_effects = fixed_effects
+  )
+  data <- enquo(data)
+  model_data <- clean_model_data(data = data, datargs)
+
+  fes <- is.character(model_data[["fixed_effects"]])
+  if (fes) {
+    yoriginal <- model_data[["outcome"]]
+    Xoriginal <- model_data[["design_matrix"]]
+    model_data <- demean_fes(model_data)
+    attr(model_data$fixed_effects, "fe_rank") <- sum(model_data[["fe_levels"]]) + 1
+  } else {
+    Xoriginal <- NULL
+    yoriginal <- NULL
+  }
+
+  return_list <-
+    lm_robust_fit(
+      y = model_data$outcome,
+      X = model_data$design_matrix,
+      yoriginal = yoriginal,
+      Xoriginal = Xoriginal,
+      weights = model_data$weights,
+      cluster = model_data$cluster,
+      fixed_effects = model_data$fixed_effects,
+      ci = ci,
+      se_type = se_type,
+      alpha = alpha,
+      return_vcov = return_vcov,
+      try_cholesky = try_cholesky,
+      has_int = attr(model_data$terms, "intercept"),
+      iv_stage = list(0)
+    )
+
+  return_lm_robust <- lm_return(
+    return_list,
+    model_data = model_data,
+    formula = formula
+  )
+
+
+  model <- return_lm_robust
+
+  out <- car::linearHypothesis(model, linearHypothesis,
+                               level = 1-alpha)
+
+  estimate  <- drop(attr(out, "value"))
+  std.error <- sqrt(diag(attr(out, "vcov")))
+  df <- model$df.residual
+  tt <- estimate/std.error
+  p.value <-  2 * pt(abs(tt), df)
+  alpha_ <- alpha/2
+  alpha_ <- c(alpha_, 1 - alpha_)
+  ci <- estimate + std.error %o% qt(alpha_, df)
+  if(is.null(lh_term)) lh_term <- linearHypothesis
+
+  return_lh <- list(   coefficients =  estimate,
```

Comment:
Why a list here and not a data.frame ?

Reply:
output didn't fit in a data.frame due to linear hypothesis object. Now it's is a returned as an attribute.